src/wp-includes/block-supports/elements.php

```diff
@@ -41,42 +41,42 @@ function wp_render_elements_support( $block_content, $block ) {
 		'button'  => array(
 			'skip'  => wp_should_skip_block_supports_serialization( $block_type, 'color', 'button' ),
 			'paths' => array(
-				'style.elements.button.color.text',
-				'style.elements.button.color.background',
-				'style.elements.button.color.gradient',
+				array( 'button', 'color', 'text' ),
```

Comment:
@dmsnell To avoid creating the arrays unnecessarily on each call, what do you think about moving the early return on lines 84-90 up to line 39 and changing them to this?

```php
$skip_button_color  = wp_should_skip_block_supports_serialization( $block_type, 'color', 'button' );
$skip_link_color    = wp_should_skip_block_supports_serialization( $block_type, 'color', 'link' );
$skip_heading_color = wp_should_skip_block_supports_serialization( $block_type, 'color', 'heading' );

if ( $skip_button_color && $skip_link_color && $skip_heading_color) {
    return $block_content;
}
```
Then this will further reduce memory usage while having no functional or visual impact.

The `$skip_*` variables _could_ be used for the `'skip'` keys in the arrays below to avoid duplicate calls, but see the comment below for an alternative approach.

Reply:
thanks @costdev - I explored this route but found diminishing returns. there are more optimization opportunities involving bailing out in a hierarchical manner too, such as saying, if there's no `style.elements.heading` attribute then don't check `style.elements.heading.color.text` but in my tests these all created marginal returns while making the code harder to understand or modify.

since these are all static we could eliminate the arrays entirely and fall back to static control structures.

again, things that I looked at but found no evidence that they are noteworthy optimizations, and things I felt could be addressed in a more normal development cycle where bigger changes can be examined.

Reply:
not sure where you all stand on adding these extra changes. I will try and create a new commit for them, but given the goals of this change and the timeline and that it's code coming in from a Gutenberg package I'm in favor of leaving this patch as-is, especially if we can't show a meaningful further improvement.